canopen/source/co_nmt.c

```diff
@@ -138,6 +138,9 @@ void CONmtSetMode(CO_NMT *nmt, CO_MODE mode)
         if (mode == CO_OPERATIONAL) {
             COTPdoInit(nmt->Node->TPdo, nmt->Node);
             CORPdoInit(nmt->Node->RPdo, nmt->Node);
+        } else if (mode == CO_PREOP) {
+            /* Activate SYNC producer if possible */
+            COSyncProdActivate(&nmt->Node->Sync);
```

Comment:
Can we move this to `COSyncInit()`?
(I'm not sure. I guess this could be possible when we consider the change regarding allowed messages in dependence to operating mode in `COSyncProdSend()`).

Reason: This would keep the SYNC functionality decoupled from NMT member functions.

Reply:
This is the only suggestion i dont know what to do about :) SYNC producer activation has to be triggered somewhere autonomously. Fieldbus processing thread is triggered only by received frames, timer thread has nothing to do with it as the SYNC producer timer is not yet existing. So the only trigger left is an user interaction with stack. I think SYNC producer activation should be hidden from user as it expects it to activate automatically based on dictionary configuration.

Reply:
PDO API is also not decoupled from netzwerkmanagement by calling `COTPdoInit()` and `CORPdoInit()` upon operational entry.

Reply:
Hmm, ok. My hope was, that a call of COSyncProdActivate() within COSyncInit() independent of NMT mode is sufficient. But when you think here is the best place: ok, accepted.

Reply:
Oh i understand now. I think its doable but in some situations it can happen that SYNC frames will be sent during NMT_INIT state. By the way, this reminds me, i was not able to find the behavior of the stack during transition to NMT_STOPPED mode, where everything except NMT is not allowed (which means stopping SYNC producer, PDOs and both SDO server and client). I will dive into it further next week.

Reply:
I think it is safe, because with our check `if ((nmt->Allowed & CO_SYNC_ALLOWED) == 0) { return }` we avoid sending in all not allowed NMT modes, regardless how we enter the mode (via CAN NMT or application API function).

Reply:
You got me on this one, it makes sense now. 👍